### PR TITLE
Fix SliverAppBar Stretch Not Triggering in NestedScrollView

### PR DESCRIPTION
**Pull request by shaochenguang**
When the scrollable widget in the NestedScrollView body has its physics set to ClampingScrollPhysics, pulling down the page does not trigger the SliverAppBar stretch effect. After the nested scrollable widget handles the scroll delta, the NestedScrollView should also handle the scroll delta again.
Here is the effect after the fix：

https://github.com/flutter/flutter/assets/33643378/b0f041f9-72be-448f-9c6f-fd2066a8be9a

https://github.com/flutter/flutter/issues/149569
https://github.com/flutter/flutter/issues/54059

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

### CODE REVIEW

**Comment by shaochenguang** [comment]
Hi @Piinks could you please help review this PR?  Thank you.

**Review by Piinks — COMMENTED** [review]
Hi @shaochenguang thanks for contributing! Can you add some tests that verify this works as expected - and so that we do not regress it in the future?

**Review by Piinks — COMMENTED** [review]
It looks like in the video that this does enable the stretch, but it jumps back when released pretty abruptly instead of smoothly sliding back.

I wasn't able to get this to work locally, can you share the code from the video?

**Comment by shaochenguang** [comment]
Hi @Piinks. I have fixed the issue mentioned above. Could you please review it again. Thank you.
https://github.com/flutter/flutter/assets/33643378/87f1502e-d00e-4021-8260-57e0bf7e05cd

File: packages/flutter/test/widgets/nested_scroll_view_test.dart [hunk 1]
```diff
@@ -3331,6 +3331,62 @@ void main() {
       areCreateAndDispose,
     );
   });
+
+  group('SliverAppBar.stretch in NestedScrollView', () {
```
Comment (Piinks) [inline packages/flutter/test/widgets/nested_scroll_view_test.dart:3335]: Is this group needed? It looks like there is only one test.
More tests would probably be good to have though. Looking at the existing SliverAppBar.stretch tests may indicate some cases this should cover.

File: packages/flutter/test/widgets/nested_scroll_view_test.dart [hunk 2]
```diff
@@ -3331,6 +3331,62 @@ void main() {
...
+      // Check if the AppBar is stretched
+      expect(appBarHeight(tester), greaterThan(expandedAppBarHeight));
+      // Check if there is any offset in the OuterController
+      expect(nestedScrollView.currentState?.outerController.offset, lessThan(0));
+
+      await tester.pumpAndSettle();
+      expect(appBarHeight(tester), expandedAppBarHeight);
+      expect(nestedScrollView.currentState?.outerController.offset, 0);
+    });
+  });
```
Comment (Piinks) [inline packages/flutter/test/widgets/nested_scroll_view_test.dart:3389]: Different platforms have default scroll physics. Can you add `variant: TargetPlatformVariant.all()` so we know this test works on all of them?

**Review by Piinks — COMMENTED** [review]
So it looks like the scroll physics must be configured a certain way for this to work. With bouncing scroll physics set for NestedScrollView.physics, and ClampingScrollPhysics set for the body's scroll view. The inline docs on NestedScrollView mention the stretch case, can you update it?

File: packages/flutter/test/widgets/nested_scroll_view_test.dart (hunk 2 above)
Comment (shaochenguang) [reply packages/flutter/test/widgets/nested_scroll_view_test.dart:3389]: done

File: packages/flutter/test/widgets/nested_scroll_view_test.dart (hunk 1 above)
Comment (shaochenguang) [reply packages/flutter/test/widgets/nested_scroll_view_test.dart:3335]: I didn't perform any previous tests on SliverAppBar.stretch, so I'm creating a test group here. Although there is only one test currently, I believe that future tests related to SliverAppBar.stretch can be added to this group.

**Comment by shaochenguang** [comment]
> So it looks like the scroll physics must be configured a certain way for this to work. With bouncing scroll physics set for NestedScrollView.physics, and ClampingScrollPhysics set for the body's scroll view. The inline docs on NestedScrollView mention the stretch case, can you update it?

Done. Please review again. Thank you

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 3]
```diff
@@ -162,7 +162,11 @@ typedef NestedScrollViewHeaderSliversBuilder = List<Widget> Function(BuildContex
 ///
 // See https://github.com/flutter/flutter/issues/54059
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart:163]: Why would this issue still be here? Isn't this PR meant to resolve the issue?

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 4]
```diff
@@ -162,7 +162,11 @@ typedef NestedScrollViewHeaderSliversBuilder = List<Widget> Function(BuildContex
...
 /// Currently, [NestedScrollView] does not support stretching the outer
-/// scrollable, e.g. when using [SliverAppBar.stretch].
+/// scrollable by default, e.g. when using [SliverAppBar.stretch].
+///
+/// If need to support stretching the outer scrollable, you need to set the
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart:167]: We should not use "you" in our docs, https://github.com/flutter/flutter/blob/master/docs/contributing/Style-guide-for-Flutter-repo.md#use-the-passive-voice-recommend-do-not-require-never-say-things-are-simple

File: packages/flutter/test/widgets/nested_scroll_view_test.dart (hunk 1 above)
Comment (Piinks) [reply packages/flutter/test/widgets/nested_scroll_view_test.dart:3335]: There are many tests for SliverAppBar.stretch in https://github.com/flutter/flutter/blob/master/packages/flutter/test/widgets/slivers_appbar_stretch_test.dart

We should make sure those scenarios work in the NestedScrollView with SliverAppBar.stretch in the tests here.

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 5]
```diff
@@ -162,7 +162,11 @@ typedef NestedScrollViewHeaderSliversBuilder = List<Widget> Function(BuildContex
...
+/// bouncing scroll physics to NestedScrollView.physics, and set
+/// ClampingScrollPhysics to the body's scroll view.
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart:169]: There should also be some sample code, like the others have for various SliverAppBar effects above.

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart (hunk 3 above)
Comment (shaochenguang) [reply packages/flutter/lib/src/widgets/nested_scroll_view.dart]: Sorry, I missed this place. I have already deleted it now.

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart (hunk 4 above)
Comment (shaochenguang) [reply packages/flutter/lib/src/widgets/nested_scroll_view.dart]: Done

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart (hunk 5 above)
Comment (shaochenguang) [reply packages/flutter/lib/src/widgets/nested_scroll_view.dart:168]: Thank you for the reminder. I have added sample code in nested_scroll_view.3.dart and moved the test to slivers.appbar_stretch_test.dart.

File: packages/flutter/lib/src/widgets/nested_scroll_view.dart [hunk 6]
```diff
@@ -160,9 +160,20 @@ typedef NestedScrollViewHeaderSliversBuilder = List<Widget> Function(BuildContex
 ///
 /// ### Stretching [SliverAppBar]s
 ///
-// See https://github.com/flutter/flutter/issues/54059
 /// Currently, [NestedScrollView] does not support stretching the outer
```
Comment (Piinks) [inline packages/flutter/lib/src/widgets/nested_scroll_view.dart:163]: This still says stretch is not supported.

Comment (shaochenguang) [reply packages/flutter/lib/src/widgets/nested_scroll_view.dart]: Done

**Comment by fmaerkl-sw** [comment]
Things that do not seem to work with this patch yet:
- Bounce at the bottom
- Bounce on "flinging" (scroll quickly, then let go and let it scroll and bounce by itself)

**Comment by Piinks** [comment]
> Things that do not seem to work with this patch yet:
> 
> * Bounce at the bottom
> * Bounce on "flinging" (scroll quickly, then let go and let it scroll and bounce by itself)

Thanks very much @fmaerkl-sw for finding these issues! @shaochenguang are these things we can fix here?

**Comment by shaochenguang** [comment]
> > Things that do not seem to work with this patch yet:
> > 
> > * Bounce at the bottom
> > * Bounce on "flinging" (scroll quickly, then let go and let it scroll and bounce by itself)
> 
> Thanks very much @fmaerkl-sw for finding these issues! @shaochenguang are these things we can fix here?

I think this issue can be resolved by creating a new pull request. What do you think? (It has been taking a long time.)

**Comment by Piinks** [comment]
> I think this issue can be resolved by creating a new pull request. What do you think? (It has been taking a long time.)

We do not land partial fixes. Can you address the issues that have been identified in this PR?

**Comment by goderbauer** [comment]
(triage): I am going to close this based on the discussion above.

**Comment by Hao-yiwen** [comment]
Can this problem be fixed?

**Pull request closed without merging**